### PR TITLE
[MRG] Better path handling

### PR DESCRIPTION
**Pull request by Titan-C**
This replaces #133 
As brought back to life by #186, Sphinx-Gallery does not handle the cases when the documentation structure is in a different place than the location of the makefile and the conf.py files.

This PR aims to fix that and preceding issues stated in  #40 #50 #115 #119 #120

For test completeness I know build in circle CI the documentation in various location, with different commands a different themes.

As can be seen from the build logs(https://circleci.com/gh/Titan-C/sphinx-gallery/12?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link). The examples are run only on the first build, later builds can take advantage of the cached files as they are referenced in the same locations and the output to the described location. For review. This are the 3 builds

* alabaster theme: https://12-26191147-gh.circle-artifacts.com/0/home/ubuntu/sphinx-gallery/alabaster_html/index.html
* classic theme: https://12-26191147-gh.circle-artifacts.com/0/home/ubuntu/sphinx-gallery/doc_build/html/tutorials/index.html
* read the docs theme: https://12-26191147-gh.circle-artifacts.com/0/home/ubuntu/sphinx-gallery/doc/_build/html/advanced_configuration.html#establishing-local-references-to-examples

Since it is possible I have missed cases where links/images/references/etc are not well resolved, the links above point to different places of the documentation I have checked, but please help me explore. And even better tell me if we can elaborate a way to test on this.

There is an issue with circle ci and the cached conda installation that is addressed in #189

### CODE REVIEW

File: sphinx_gallery/gen_gallery.py [hunk 1]
```diff
@@ -99,20 +100,18 @@ def generate_gallery_rst(app):
...
+    mod_examples_dir = os.path.join(
+        app.builder.srcdir, gallery_conf['mod_example_dir'])
     seen_backrefs = set()
 
     computation_times = []
 
     # cd to the appropriate directory regardless of sphinx configuration
     working_dir = os.getcwd()
-    os.chdir(app.builder.srcdir)
+    # os.chdir(app.builder.srcdir)
```
Comment (sciunto) [inline sphinx_gallery/gen_gallery.py]: You probably want to remove this. :)

**Comment by sciunto** [comment]
This patch works well for scikit-image: https://github.com/scikit-image/scikit-image/pull/2395

**Comment by sciunto** [comment]
@Titan-C Do you plan to release a new version soon? That would be helpful for scikit-image.

PS: just realized that we are in the same building :)

**Comment by Titan-C** [comment]
Yes, I do plan to release 0.1.8 soon. I'm putting the milestones now. 
I did edited quite a few times this PR last week, but then got no new reviews.  It still works. I put new links to the artifacts

* Alabaster : https://29-26191147-gh.circle-artifacts.com/0/home/ubuntu/sphinx-gallery/alabaster_html/tutorials/plot_notebook.html#sphx-glr-tutorials-plot-notebook-py
* Read the docs: https://29-26191147-gh.circle-artifacts.com/0/home/ubuntu/sphinx-gallery/doc/_build/html/auto_examples/sin_func/plot_sin_black_background.html#sphx-glr-auto-examples-sin-func-plot-sin-black-background-py
* classic: https://29-26191147-gh.circle-artifacts.com/0/home/ubuntu/sphinx-gallery/doc_build/html/auto_mayavi_examples/index.html

**Comment by lesteve** [comment]
I was on holiday around Christmas and I am currently going through the backlog. I'll look at this one today.

File: sphinx_gallery/tests/test_gen_rst.py [hunk 2]
```diff
@@ -237,25 +238,25 @@ def test_save_figures():
...
     plt.plot(1, 1)
-    fname_template = os.path.join(examples_dir, 'image{0}.png')
-    fig_list, _ = sg.save_figures(fname_template, 0, gallery_conf)
-    assert_equal(len(fig_list), 2)
-    assert fig_list[0].endswith('image1.png')
-    assert fig_list[1].endswith('image2.png')
+    fname_template = os.path.join(gallery_conf['gallery_dir'], 'image{0}.png')
+    image_rst, fig_num = sg.save_figures(fname_template, 0, gallery_conf)
+    assert_equal(fig_num, 2)
+    assert re.search('/image1.png', image_rst)
```
Comment (lesteve) [inline sphinx_gallery/tests/test_gen_rst.py]: I pushed a change to use `in` which is simpler for simple string comparisons

File: setup.cfg [hunk 3]
```diff
@@ -7,6 +7,6 @@ with-doctest=1
 doctest-extension=rst
 
 [build_sphinx]
-source-dir = docs/
-build-dir  = docs/_build
+source-dir = doc/
```
Comment (Titan-C) [inline setup.cfg]: Yes, this was always broken as it is the setup to build the docs from setuptools, and setup tools launches the build from where `setup.py` is called. So this never worked, you can even see that it has the old `docs` path, because it was never used, and thus never took care of updating.
But now that we can call sphinx-build from any place I thought to make use of this. Because if setuptools can build the docs it can also directly upload them to pythonhosted.org when uploading the package to PyPI. Of course we stopped hosting the docs in pythonhosted long ago in favor of readthedocs.
In circle.yml all 3 statements in the test do the same, is only to test the build from different places a calling from different commands.

File: sphinx_gallery/gen_gallery.py [hunk 4]
```diff
@@ -191,32 +184,34 @@ def sumarize_failing_examples(app, exception):
         return
 
     gallery_conf = app.config.sphinx_gallery_conf
-    failing_examples = set([os.path.normpath(path) for path in
-                            gallery_conf['failing_examples']])
-    expected_failing_examples = set([os.path.normpath(path) for path in
+    failing_examples = {os.path.normpath(path): traceback
```
Comment (Titan-C) [inline sphinx_gallery/gen_gallery.py]: Yes this is entirely related to the PR. The thing is now I pass the absolute path of the files during execution. But at the end when I verify which examples failed I'm matching strings containing the path and those paths must be equal and thus can't have the `../` entries that tend to appear when joining relative paths(Which is how we configure the gallery, relative to `conf.py`) so the way I found to fix this is to normalize the paths at the end.

**Comment by Titan-C** [comment]
Comments have been addressed.

File: sphinx_gallery/gen_gallery.py (hunk 4 above)
Comment (Titan-C) [inline sphinx_gallery/gen_gallery.py]: while debugging my other PR I realized I don't need this.

File: setup.cfg (hunk 3 above)
Comment (lesteve) [inline setup.cfg]: > But now that we can call sphinx-build from any place I thought to make use of this. 

I think that is what I have an issue with, `python setup.py build_sphinx` should build the doc in the same place as `cd doc && make html` (principle of least surprise). In circle.yml use explicit sphinx-build commands to test out different locations and theme.

Comment (Titan-C) [inline setup.cfg]: Ok, least surprise in the doc builds. Travis will do the `cd doc && make html` to build from the Makefile path. In CircleCI `python setup.py build_sphinx` will do it from the project root but build in our default location `doc/_build/html`, additionally `sphinx-build is called for 2 different themes that output in different directories.

File: sphinx_gallery/gen_gallery.py (hunk 4 above)
Comment (lesteve) [inline sphinx_gallery/gen_gallery.py]: > while debugging my other PR I realized I don't need this.

Can you remove this change if it is not needed?

File: sphinx_gallery/gen_gallery.py [hunk 5]
```diff
@@ -86,6 +86,7 @@ def generate_gallery_rst(app):
         abort_on_example_error=app.builder.config.abort_on_example_error)
 
     # this assures I can call the config in other places
```
Comment (lesteve) [inline sphinx_gallery/gen_gallery.py]: This comment is related to the line `app.config.sphinx_gallery_conf = gallery_conf`, isn't it?

**Comment by Titan-C** [comment]
> This comment is related to the line `app.config.sphinx_gallery_conf = gallery_conf`, isn't it?

Yes, I'll move it

**Comment by Titan-C** [comment]
Is already done, I removed it as soon I realized it was not needed.

**Comment by Titan-C** [comment]
I just realized a last minute bug thanks to @chsasank, the zip downloads were using the wrong path. Now it is good.

File: sphinx_gallery/gen_gallery.py [hunk 6]
```diff
@@ -191,32 +184,33 @@ def sumarize_failing_examples(app, exception):
         return
 
     gallery_conf = app.config.sphinx_gallery_conf
-    failing_examples = set([os.path.normpath(path) for path in
-                            gallery_conf['failing_examples']])
-    expected_failing_examples = set([os.path.normpath(path) for path in
+    failing_examples = gallery_conf['failing_examples']
```
Comment (lesteve) [inline sphinx_gallery/gen_gallery.py]: I don't think you reverted all the unnecessary changes. Keep failing_examples as it was previously.

Keep the `\n\nExamples` if you wish.

**Comment by Titan-C** [comment]
> I don't think you reverted all the unnecessary changes. Keep failing_examples as it was previously.

Changed to disturb the least possible previous code in
`summarize_failing_examples`, changed required for the PR are there.

**Comment by lesteve** [comment]
Just noticed the `orphan:` in https://49-26191147-gh.circle-artifacts.com/0/home/ubuntu/sphinx-gallery/doc/_build/html/auto_examples/index.html.

Is there a good reason why it is here?

**Comment by Titan-C** [comment]
> Just noticed the `orphan:` in https://49-26191147-gh.circle-artifacts.com/0/home/ubuntu/sphinx-gallery/doc/_build/html/auto_examples/index.html.
>
> Is there a good reason why it is here?

I have noticed it too some time long ago. It was introduced in PR #187,
and I have fixed in older PRs to be merged #160 and #151. But just
forgot to fix it here too, it was not really the aim of the PR.

**Comment by Titan-C** [comment]
I fixed the `orphan` in PR#198

**Comment by lesteve** [comment]
> I fixed the `orphan` in PR #198

Nice, thanks for doing that in a separate PR!

**Comment by lesteve** [comment]
LGTM, I pushed some cosmetic changes. I'll merge this if this is green.

**Comment by lesteve** [comment]
Green, merging, thanks a lot this is some very nice work to fix in a neat way a long-standing bug!
